fullrt/dht.go

```diff
@@ -963,83 +954,209 @@ func (dht *FullRT) PutMany(ctx context.Context, keys []string, values [][]byte)
-		}, peers, true)
-		if successes == 0 {
-			return fmt.Errorf("no successful puts")
-		}
-		return nil
+	fn := func(ctx context.Context, p, k peer.ID) error {
+		keyStr := string(k)
+		return dht.protoMessenger.PutValue(ctx, p, record.MakePutRecord(keyStr, keyRecMap[keyStr]))
 	}
 
 	return dht.bulkMessageSend(ctx, keysAsPeerIDs, fn, false)
 }
 
-func (dht *FullRT) bulkMessageSend(ctx context.Context, keys []peer.ID, fn func(ctx context.Context, k peer.ID) error, isProvRec bool) error {
+func (dht *FullRT) bulkMessageSend(ctx context.Context, keys []peer.ID, fn func(ctx context.Context, target, k peer.ID) error, isProvRec bool) error {
 	if len(keys) == 0 {
 		return nil
 	}
 
-	sortedKeys := kb.SortClosestPeers(keys, kb.ID(make([]byte, 32)))
+	type report struct {
+		successes   int
+		failures    int
+		lastSuccess time.Time
+	}
+
+	successMapLk := sync.RWMutex{}
+	keySuccesses := make(map[peer.ID]*report)
+	numSkipped := 0
+
+	for _, k := range keys {
+		keySuccesses[k] = &report{}
+	}
 
-	var numSends uint64 = 0
-	var numSendsSuccessful uint64 = 0
+	logger.Infof("bulk send: number of keys %d, unique %d", len(keys), len(keySuccesses))
+	numSuccessfulToWaitFor := int(float64(dht.bucketSize) * dht.waitFrac * 1.2)
 
+	sortedKeys := make([]peer.ID, 0, len(keySuccesses))
+	for k := range keySuccesses {
+		sortedKeys = append(sortedKeys, k)
+	}
+
+	sortedKeys = kb.SortClosestPeers(sortedKeys, kb.ID(make([]byte, 32)))
+
+	dht.kMapLk.RLock()
+	numPeers := len(dht.keyToPeerMap)
+	dht.kMapLk.RUnlock()
+
+	chunkSize := (len(sortedKeys) * dht.bucketSize * 2) / numPeers
+	if chunkSize == 0 {
+		chunkSize = 1
+	}
+
+	connmgrTag := fmt.Sprintf("dht-bulk-provide-tag-%d", rand.Int())
+
+	type workMessage struct {
+		p    peer.ID
+		keys []peer.ID
+	}
+
+	workCh := make(chan workMessage, 1)
 	wg := sync.WaitGroup{}
-	onePctKeys := uint64(len(sortedKeys)) / 100
+	wg.Add(dht.bulkSendParallelism)
+	for i := 0; i < dht.bulkSendParallelism; i++ {
+		go func() {
+			defer wg.Done()
+			defer logger.Debugf("bulk send goroutine done")
+			for {
+				select {
+				case wmsg, ok := <-workCh:
+					if !ok {
+						return
+					}
+					p, workKeys := wmsg.p, wmsg.keys
+					dht.peerAddrsLk.RLock()
+					peerAddrs := dht.peerAddrs[p]
+					dht.peerAddrsLk.RUnlock()
+					dialCtx, dialCancel := context.WithTimeout(ctx, dht.timeoutPerOp)
+					if err := dht.h.Connect(dialCtx, peer.AddrInfo{ID: p, Addrs: peerAddrs}); err != nil {
+						dialCancel()
+						successMapLk.Lock()
+						numSkipped++
+						successMapLk.Unlock()
+						continue
+					}
+					dialCancel()
+					dht.h.ConnManager().Protect(p, connmgrTag)
+					for _, k := range workKeys {
+						successMapLk.RLock()
+						keyReport := *keySuccesses[k]
+						successMapLk.RUnlock()
+
+						queryTimeout := dht.timeoutPerOp
+						if keyReport.successes >= numSuccessfulToWaitFor {
+							if time.Since(keyReport.lastSuccess) > time.Millisecond*500 {
+								continue
```

Comment:
Does it make sense to add a bool `keyDone` to the `keyReport` and set it to `true` here?  This would prevent waiting again if a straggler response comes in and updates the `lastSuccessTime`.  Also, for keys that have enough success and have already waited 500ms, this would avoid another system call to get time.